demos/http/http_demo_s3_download/http_demo_s3_download.c

```diff
@@ -0,0 +1,625 @@
+#ifndef TRANSPORT_SEND_RECV_TIMEOUT_MS
+    #define TRANSPORT_SEND_RECV_TIMEOUT_MS    ( 1000 )
+#endif
+
+/* Check that size of the user buffer is defined. */
+#ifndef USER_BUFFER_LENGTH
+    #define USER_BUFFER_LENGTH    ( 4096 )
+#endif
+
+/* Check that size of the file download buffer is defined. */
+#ifndef FILE_BUFFER_LENGTH
+    #define FILE_BUFFER_LENGTH    ( 2048 )
+#endif
+
+/**
+ * @brief The length of the HTTP GET method.
+ */
+#define HTTP_METHOD_GET_LENGTH                    ( sizeof( HTTP_METHOD_GET ) - 1 )
+
+/**
+ * @brief Field name of the HTTP Range header to read from server response.
+ */
+#define HTTP_CONTENT_RANGE_HEADER_FIELD           "Content-Range"
+
+/**
+ * @brief Length of the HTTP Range header field.
+ */
+#define HTTP_CONTENT_RANGE_HEADER_FIELD_LENGTH    ( sizeof( HTTP_CONTENT_RANGE_HEADER_FIELD ) - 1 )
+
+/**
+ * @brief A buffer used in the demo for storing HTTP request headers and
+ * HTTP response headers and body.
+ *
+ * @note This demo shows how the same buffer can be re-used for storing the HTTP
+ * response after the HTTP request is sent out. Here, a separate buffer is used
+ * to download the file from the server. However, the user can decide how to use
+ * buffers to store HTTP requests and responses.
+ */
+static uint8_t userBuffer[ USER_BUFFER_LENGTH ];
+
+/**
+ * @brief A buffer used in the demo for downloading the file from the server
+ * in chunks.
+ */
+static uint8_t fileDownloadBuffer[ FILE_BUFFER_LENGTH ];
+
+/**
+ * @brief Represents header data that will be sent in an HTTP request.
+ */
+static HTTPRequestHeaders_t requestHeaders;
+
+/**
+ * @brief Configurations of the initial request headers that are passed to
+ * #HTTPClient_InitializeRequestHeaders.
+ */
+static HTTPRequestInfo_t requestInfo;
+
+/**
+ * @brief Represents a response returned from an HTTP server.
+ */
+static HTTPResponse_t response;
+
+/* The location of the host address within string S3_PRESIGNED_URL. */
+static const char * pAddress = NULL;
+
+/* The host address string extracted from S3_PRESIGNED_URL. */
+static char serverHost;
+
+/* The length of the host address found in string S3_PRESIGNED_URL. */
+static size_t serverHostLength = 0;
+
+/*-----------------------------------------------------------*/
+
+/**
+ * @brief Connect to HTTP server with reconnection retries.
+ *
+ * @param[out] pNetworkContext The output parameter to return the created network context.
+ *
+ * @return EXIT_FAILURE on failure; EXIT_SUCCESS on successful connection.
+ */
+static int32_t connectToServer( NetworkContext_t * pNetworkContext );
+
+/**
+ * @brief Retrieve the size of the S3 object that is specified in pPath.
+ *
+ * @param[out] pFileSize - The size of the S3 object.
+ * @param[in] pTransportInterface The transport interface for making network calls.
+ * @param[in] pHost The host name of the server.
+ * @param[in] pMethod The HTTP request method.
+ * @param[in] pPath The Request-URI to the objects of interest.
```

Comment:
Can you note that these are NULL terminated strings?
```suggestion
 * @param[in] pHost The host name of the server. This string must be NULL terminated.
 * @param[in] pMethod The HTTP request method. This string must be NULL terminated.
 * @param[in] pPath The Request-URI to the objects of interest. This string must be NULL terminated.
```